### PR TITLE
Add notifications for statuses deleted by moderators

### PR DESCRIPTION
**Pull request by Gargron**
Redesign the report screen in the admin UI, add categorization to reports, allow moderators to change which statuses are in the report, include categorization in e-mails to the reported account, and make deleting offending content a viable moderation option by notifying the reported account about it instead of doing it silently.

The redesign serves multiple goals:

1. Make the UI more obvious with a better logical flow and more hints
2. Reduce necessary clicks to different pages by displaying more at-a-glance information and actions
3. Make it easier to spot the difference between real and spam accounts when they get reported
4. Make it easier to prioritize reports in the queue but also improve communication to the reported user with categories

Besides updates to the page structure, the redesign includes a new look for report notes and an improved look for the audit log. The ability to silently delete statuses outside of reports in the admin UI has been replaced by the ability to create reports for those statuses right from the admin UI. While this introduces an extra step, this improves accountability in the system.

Moderators have the ability to edit the category of the report, and add and remove statuses from the report, allowing them to fix often incomplete reports before proceeding to take action, resulting in better record-keeping and more accurate communication with the reported account.

Each punitive action against an account, such as deleting a set of statuses or a suspension, is additionally recorded as a "strike". The number of previous strikes is highlighted on the report screen, allowing moderators to escalate punitive measures for repeat offenders without having to navigate to the audit log or past reports.

image

> Note: The report in the above screenshot is an account reporting itself.

image

### Report categories

This PR introduces the concept of a report category, which can be "Other" (all legacy reports, and whatever doesn't fit in the other categories), "Spam", and "Rule violation". "Rule violation" allows selecting which of the server rules that are defined in the system have been violated. This allows two things:

1. Prioritizing the handling of more severe reports in the queue before others
2. Give the reported account more details about the reason an action has been taken against them

___

Funded through the BMBF (BMBF Förderkennzeichen: 01IS21S29)

### CODE REVIEW

**Comment by ClearlyClaire** [comment]
I have not reviewed any of the code yet, but based on the description and screenshots:
- I'm not sure requiring a mail to be sent to the reported user on toot deletion is a good idea, but having the ability to send such a mail is definitely a good idea and a good default
- the report page looks a lot nicer, key info is easier to see, and common actions being readily available is definitely a good thing

However, I'm a bit worried about:
- “Delete and resolve” making it awkward to, for instance, delete the toots *and* limit the reported user
- “Add more to report” and “Remove from report” potentially muddying the report's context, especially when multiple moderators are involved, and for instance allowing a malicious moderator to swap statuses to discredit a reporter

Minor improvement ideas that can definitely wait for another PR:
- in the “Reported by” line, display the number of reports the reporter has made
- Along with “Previous strikes”, display the amount of open reports targeting the reported user
- Make “Previous strikes” link to those previous strikes

Another related idea that can definitely wait for another PR: offer to automatically close other reports targeting the same user if they have the same category and target only a (non-empty) subset of the same toots.

File: app/views/admin/reports/show.html.haml
Comment (ClearlyClaire) [inline app/views/admin/reports/show.html.haml:60]: Would be useful to link to those strikes and also have quick access to other currently open reports about that user.

File: app/views/admin/reports/show.html.haml
Comment (ClearlyClaire) [inline app/views/admin/reports/show.html.haml:75]: A useful addition might be the number of reports from that user, with a quick link to see them.

File: app/views/admin/reports/show.html.haml
Comment (ClearlyClaire) [inline app/views/admin/reports/show.html.haml:122]: Very minor concern about requiring javascript here, which I think is new for most of the admin UI (there was the dashboard but it's less critical).

File: app/views/admin/reports/show.html.haml
Comment (ClearlyClaire) [inline app/views/admin/reports/show.html.haml:204]: Notes and action logs being displayed in the same list made sense, and I would prefer if we found a design that worked in the same way.

File: app/workers/scheduler/user_cleanup_scheduler.rb
Comment (ClearlyClaire) [inline app/workers/scheduler/user_cleanup_scheduler.rb:33]: Naive question: is this efficient?

File: db/migrate/20211231080958_add_category_to_reports.rb
Comment (ClearlyClaire) [inline db/migrate/20211231080958_add_category_to_reports.rb:18]: I'd add something to set `action_taken` whenever `action_taken_at` is non-null.

File: app/models/appeal.rb
Comment (ClearlyClaire) [inline app/models/appeal.rb:16]: Shouldn't that be `account_warning`? More generally, we aren't doing anything with appeals and they are not mentioned in the PR description, I'd just remove them from the current PR.

File: spec/fabricators/account_strike_fabricator.rb
Comment (ClearlyClaire) [inline spec/fabricators/account_strike_fabricator.rb:1]: Should that exist? Isn't it `account_warning`?

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
## Undocumented changes

There are a few things that have been changed by this PR without being documented in the description:
- the ability for moderators to “Mark [a toot] as sensitive” has been removed, I'm not sure how much that is used in practice, and the use is limited for local accounts as that won't push an update to the network, but I think this is sometimes used by moderators on remote users
- database changes and models have been added for an “appeal” system; that's not an issue, but since there is no code to do anything with it yet, I think this should be omitted from the current PR

## Various design remarks

From what I perceive as the most important to the least important:
- the ability for mods to add and remove statuses from reports can be confusing, I would greatly prefer if those would be different sections (e.g. “reported statuses judged as irrelevant”, “additional statuses added by mods”), so as to keep the context intact and improve accountability
- rules might change after a report got created, which may make the report misleading
- I'm not sure we need the hardcoded categories… I'm afraid there would be significant overlap with the admin-defined server rules… the only benefit I see is translation of those categories, which makes me think we should make server rules translatable, but that's for another PR
- Tying deleting toots with closing the report in “Delete and resolve” makes some things awkward and requiring two strikes for only one offense (e.g. if you want to delete some of the offending toots and also limit the user)

File: app/models/account_warning.rb
Comment (Gargron) [reply app/models/account_warning.rb:24]: No, they are updated in a migration. It is to prevent an event like when the force-sensitive option was added shifting existing records' enums by leaving enough space between values to insert new types.

File: app/models/report.rb
Comment (Gargron) [reply app/models/report.rb:46]: To be able to insert more enum types in the future. It's an int column anyway, there is no benefit from having smaller values.

File: app/models/report_filter.rb
Comment (Gargron) [reply app/models/report_filter.rb:22]: Because the base scope is `Report.unresolved`, this is needed for `Report.resolved` to actually work when merged in, as per deprecation warning.

File: app/views/admin/report_notes/_report_note.html.haml
Comment (Gargron) [reply app/views/admin/report_notes/_report_note.html.haml:11]: Intentional UI choice.

File: app/views/admin/reports/show.html.haml
Comment (Gargron) [reply app/views/admin/reports/show.html.haml:75]: Never found that information useful for any decisions.

File: app/models/admin/status_batch_action.rb
Comment (Gargron) [reply app/models/admin/status_batch_action.rb:75]: Possible future improvement but feel like it would be premature right now.

File: app/controllers/api/v1/admin/account_actions_controller.rb
Comment (Gargron) [reply app/controllers/api/v1/admin/account_actions_controller.rb:6]: Allows cookie-authenticated requests to this API

File: app/controllers/admin/report_notes_controller.rb
Comment (Gargron) [reply app/controllers/admin/report_notes_controller.rb:29]: To show validation errors within the form.

**Comment by Gargron** [comment]
Removed appeals stubs from PR.

File: app/workers/scheduler/user_cleanup_scheduler.rb
Comment (Gargron) [reply app/workers/scheduler/user_cleanup_scheduler.rb:33]: Valid point, this is not at all efficient, looks like it's a seq scan.

**Comment by Gargron** [comment]
Added index on `deleted_at` on the `statuses` table.

File: app/models/account_warning.rb
Comment (ClearlyClaire) [reply app/models/account_warning.rb:24]: But inserting new types at the end isn't an issue, is it?

File: app/models/report.rb
Comment (ClearlyClaire) [reply app/models/report.rb:46]: Yeah it's not bad, just weird.

File: app/models/admin/status_batch_action.rb
Comment (ClearlyClaire) [reply app/models/admin/status_batch_action.rb:75]: That can wait for a future PR but I really want to have something like this before these changes get released. It's not a very involved change either.

File: spec/fabricators/appeal_fabricator.rb
Comment (ClearlyClaire) [inline spec/fabricators/appeal_fabricator.rb:5]: This file should be deleted.

File: app/models/account_warning.rb
Comment (ClearlyClaire) [reply app/models/account_warning.rb:24]: Not using these weird values would mean we can skip the migration. Not a huge thing, but still.

File: app/models/report.rb
Comment (ClearlyClaire) [inline app/models/report.rb:115]: These include actions that date from before the report, I don't know whether that's deliberate or not. It can be useful, but also noisy or confusing.

File: app/views/admin/report_notes/_report_note.html.haml
Comment (ClearlyClaire) [reply app/views/admin/report_notes/_report_note.html.haml:11]: Oh, nevermind, I misread this, I thought it only displayed “Today”, but it's “Today at …”

File: app/models/report.rb
Comment (Gargron) [reply app/models/report.rb:115]: Deliberate, I think it's more useful to see an account has been acted on before...

Comment (ClearlyClaire) [reply app/models/report.rb:115]: I think it can be confusing but it's probably not a big deal. We'll see I guess.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
My following concerns remain, and I expect them to be addressed before release, but I think they don't need to block this PR:
- the PR removes the ability for moderators to “Mark [a toot] as sensitive”
- toots added/removed to an existing report by mods should be clearly stated as such, to avoid altering the context of the original report
- to a lesser extent, deleting toots with closing the report in “Delete and resolve” makes some things awkward, requires two strikes, and the order in which actions are taken greatly change the flow (e.g. deleting first is probably preferable, as silencing first will close reports and you'd have to search for them again)
- also to a lesser extent, I think more consideration may be needed with regards to hardcoded categories, as I feel they may end up either partially redundant with possibly more detailed server rules, or not as universal as planned

One thing that I'd like to see changed before the PR is merged, although not a hard requirement, is not renumbering `account_warnings.action` as to avoid an unnecessary migration step.

**Comment by Gargron** [comment]
> One thing that I'd like to see changed before the PR is merged, although not a hard requirement, is not renumbering account_warnings.action as to avoid an unnecessary migration step.

It’s on a tiny table and it’s just better if the numbers correspond to the severity of the action. It feels weird if delete_statuses was 5 and suspend was 4.

**Comment by ClearlyClaire** [comment]
> > One thing that I'd like to see changed before the PR is merged, although not a hard requirement, is not renumbering account_warnings.action as to avoid an unnecessary migration step.
> 
> It’s on a tiny table and it’s just better if the numbers correspond to the severity of the action. It feels weird if delete_statuses was 5 and suspend was 4.

Alright :woman_shrugging:

**Pull request merged**
